File: apps/windows/SunoFlow/BrandMark.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Drawing2D;

namespace SunoFlow;

/// <summary>
/// The SunoFlow brand mark — an ear receiving sound.
///
/// The path strings below are copied verbatim from the website's wordmark
/// (<c>site/assets/favicon.svg</c>), so the tray, the dashboard, the app icon and
/// the web app cannot drift apart. They are SVG path data in a 24x24 box, stroked
/// at width 2 with round caps and joins. This is the C# counterpart of
/// <c>SunoFlowApp/Sources/SunoFlow/BrandMark.swift</c> and of the rasteriser in
/// <c>tools/make-icons.js</c> — three renderers, one geometry.
///
/// Unlike Quartz, GDI+ already grows y downwards, so no flip is needed here.
/// </summary>
internal static class BrandMark
{
    /// <summary>The outer ear. A closed subpath, so it can also be filled as a silhouette.</summary>
    public const string EarPath =
        "M8.9 4.3c2.8 0 4.7 2 4.7 4.8 0 2.5-1.8 3.7-2.9 5.1-.7.9-.9 1.8-.9 2.9 0 1.5-1.2 2.6-2.6 2.6s-2.6-1.2-2.6-2.6c0-1.2.4-2 .4-3.1 0-1.6-1-2.8-1-4.9 0-2.8 2.1-4.8 4.9-4.8z";

    /// <summary>The inner curl of the ear.</summary>
    public const string CurlPath =
        "M8.9 7.4c1.4 0 2.2 1 2.2 2.1 0 1.3-1.1 1.9-1.7 2.8-.4.6-.5 1.2-.5 1.9";

    /// <summary>The two arcs of sound arriving at the ear.</summary>
    public static readonly string[] WavePaths =
    {
        "M16.6 10.2a3.6 3.6 0 0 0 0 5.2",
        "M19.9 8.2a6.8 6.8 0 0 0 0 9.2",
    };

    /// <summary>The design box the paths live in, and the weight they are drawn at.</summary>
    public const float Box = 24f;
    public const float Stroke = 2f;

    /// <summary>
    /// How much of the mark to draw. The ear is always the mark; what changes is
    /// whether sound is arriving (<see cref="Idle"/>), being taken in
    /// (<see cref="Recording"/> — a filled silhouette), or not being listened for
    /// at all (<see cref="Processing"/> is busy, <see cref="Offline"/> is struck
    /// through).
    /// </summary>
    public enum Variant { Idle, Recording, Processing, Offline }

    /// <summary>
    /// Draws the mark centred in <paramref name="rect"/>, scaled to the shorter side.
    /// </summary>
    /// <param name="knockout">
    /// The colour to punch the offline slash's gap in. GDI+ has no "clear" blend
    /// mode when drawing into a control, so the gap is painted rather than erased;
    /// pass the surface colour behind the mark. Ignored for every other variant.
    /// </param>
    public static void Draw(Graphics g, RectangleF rect, Color color,
                            Variant variant = Variant.Idle, Color? knockout = null)
    {
        float scale = Math.Min(rect.Width, rect.Height) / Box;
        if (scale <= 0) return;

        using var transform = new Matrix();
        transform.Translate(rect.X + (rect.Width - Box * scale) / 2f,
                            rect.Y + (rect.Height - Box * scale) / 2f);
        transform.Scale(scale, scale);

        var savedSmoothing = g.SmoothingMode;
        g.SmoothingMode = SmoothingMode.AntiAlias;
        using var pen = new Pen(color, Stroke * scale)
        {
            StartCap = LineCap.Round,
            EndCap = LineCap.Round,
            LineJoin = LineJoin.Round,
        };
        using var brush = new SolidBrush(color);

        using (var ear = Placed(EarPath, transform))
        {
            if (variant == Variant.Recording)
            {
                // Fill *and* stroke, so the silhouette keeps the same outer edge as
                // the outlined variants and the mark doesn't jump size on record.
                g.FillPath(brush, ear);
                g.DrawPath(pen, ear);
            }
            else
            {
                g.DrawPath(pen, ear);
                using var curl = Placed(CurlPath, transform);
                g.DrawPath(pen, curl);
            }
        }

        if (variant == Variant.Idle || variant == Variant.Recording)
        {
            foreach (var wave in WavePaths)
            {
                using var path = Placed(wave, transform);
                g.DrawPath(pen, path);
            }
        }

        if (variant == Variant.Offline)
        {
            var ends = new[] { new PointF(4.2f, 19.8f), new PointF(19.8f, 4.2f) };
            transform.TransformPoints(ends);
            if (knockout is Color gap)
            {
                // Knock a gap out from under the slash first, so it reads as one
                // stroke crossing the mark rather than a line tangled up in it.
                using var eraser = new Pen(gap, Stroke * scale * 2.4f)
                {
                    StartCap = LineCap.Round,
                    EndCap = LineCap.Round,
                };
                g.DrawLine(eraser, ends[0], ends[1]);
            }
            g.DrawLine(pen, ends[0], ends[1]);
        }

        g.SmoothingMode = savedSmoothing;
    }

    // MARK: - SVG path reading

    /// <summary>Parsed once; the mark is fixed at build time.</summary>
    private static readonly Dictionary<string, GraphicsPath> Cache = new();

    private static GraphicsPath Placed(string d, Matrix transform)
    {
        GraphicsPath? cached;
        lock (Cache)
        {
            if (!Cache.TryGetValue(d, out cached))
            {
                cached = Parse(d);
                Cache[d] = cached;
            }
        }
        var copy = (GraphicsPath)cached.Clone();
        copy.Transform(transform);
        return copy;
    }

    /// <summary>
    /// A deliberately small SVG path reader: it understands exactly the commands
    /// the mark uses (M, L, C, S, A, Z, and their relative forms). The alternative
    /// is hand-porting the curves into C#, where they would quietly drift away
    /// from the ones the website ships.
    /// </summary>
    private static GraphicsPath Parse(string d)
    {
        var path = new GraphicsPath();
        int i = 0;
        PointF cur = PointF.Empty, start = PointF.Empty;
        PointF? lastControl = null;
        char command = 'M';

        while (true)
        {
            SkipSeparators(d, ref i);
            if (i >= d.Length) break;
            if (char.IsLetter(d[i])) command = d[i++];
            bool relative = char.IsLower(command);

            switch (char.ToLowerInvariant(command))
            {
                case 'm':
                {
                    var p = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    path.StartFigure();
                    cur = p; start = p; lastControl = null;
                    // Extra coordinate pairs after a moveto are implicit linetos.
                    command = relative ? 'l' : 'L';
                    break;
                }
                case 'l':
                {
                    var p = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    path.AddLine(cur, p);
                    cur = p; lastControl = null;
                    break;
                }
                case 'c':
                {
                    var c1 = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    var c2 = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    var end = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    path.AddBezier(cur, c1, c2, end);
                    cur = end; lastControl = c2;
                    break;
                }
                case 's':
                {
                    // Smooth curve: the first control point mirrors the previous one.
                    var previous = lastControl ?? cur;
                    var c1 = new PointF(2 * cur.X - previous.X, 2 * cur.Y - previous.Y);
                    var c2 = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    var end = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    path.AddBezier(cur, c1, c2, end);
                    cur = end; lastControl = c2;
                    break;
                }
                case 'a':
                {
                    float rx = Number(d, ref i), ry = Number(d, ref i);
                    float rotation = Number(d, ref i);
                    bool largeArc = Number(d, ref i) != 0, sweep = Number(d, ref i) != 0;
                    var end = Resolve(cur, relative, Number(d, ref i), Number(d, ref i));
                    AddArc(path, cur, end, rx, ry, rotation, largeArc, sweep);
                    cur = end; lastControl = null;
                    break;
                }
                case 'z':
                {
                    path.CloseFigure();
                    cur = start; lastControl = null;
                    break;
                }
                default:
                    return path;   // An unsupported command: stop rather than guess.
            }
        }
        return path;
    }

    private static PointF Resolve(PointF cur, bool relative, float x, float y) =>
        relative ? new PointF(cur.X + x, cur.Y + y) : new PointF(x, y);

    private static bool IsSeparator(char c) =>
        c == ' ' || c == ',' || c == '\n' || c == '\t' || c == '\r';

    private static void SkipSeparators(string d, ref int i)
    {
        while (i < d.Length && IsSeparator(d[i])) i++;
    }

    /// <summary>Reads one number. SVG lets them run together without separators
    /// (<c>.4-3.1</c> is <c>0.4</c> then <c>-3.1</c>), so this stops at the first
    /// character that cannot continue the current number.</summary>
    private static float Number(string d, ref int i)
    {
        SkipSeparators(d, ref i);
        int begin = i;
        if (i < d.Length && (d[i] == '-' || d[i] == '+')) i++;
        bool seenDot = false;
        while (i < d.Length)
        {
            char c = d[i];
            if (c >= '0' && c <= '9') i++;
            else if (c == '.' && !seenDot) { seenDot = true; i++; }
            else break;
        }
        var text = d.Substring(begin, i - begin);
        return float.TryParse(text, System.Globalization.NumberStyles.Float,
                              System.Globalization.CultureInfo.InvariantCulture, out var value)
            ? value : 0f;
    }

    /// <summary>Endpoint-parameterised elliptical arc to cubic Béziers (SVG spec F.6.5).</summary>
    private static void AddArc(GraphicsPath path, PointF p0, PointF p1,
                               float rxIn, float ryIn, float degrees, bool largeArc, bool sweep)
    {
        if (p0 == p1) return;
        double rx = Math.Abs(rxIn), ry = Math.Abs(ryIn);
        if (rx == 0 || ry == 0) { path.AddLine(p0, p1); return; }

        double phi = degrees * Math.PI / 180.0;
        double cosPhi = Math.Cos(phi), sinPhi = Math.Sin(phi);
        double dx = (p0.X - p1.X) / 2.0, dy = (p0.Y - p1.Y) / 2.0;
        double x1 = cosPhi * dx + sinPhi * dy;
        double y1 = -sinPhi * dx + cosPhi * dy;

        // Grow the radii if they're too small to span the chord (spec F.6.6).
        double lambda = (x1 * x1) / (rx * rx) + (y1 * y1) / (ry * ry);
        if (lambda > 1) { double s = Math.Sqrt(lambda); rx *= s; ry *= s; }

        double sign = largeArc != sweep ? 1 : -1;
        double numerator = Math.Max(0, rx * rx * ry * ry - rx * rx * y1 * y1 - ry * ry * x1 * x1);
        double denominator = rx * rx * y1 * y1 + ry * ry * x1 * x1;
        double coefficient = denominator == 0 ? 0 : sign * Math.Sqrt(numerator / denominator);
        double cx1 = coefficient * rx * y1 / ry;
        double cy1 = -coefficient * ry * x1 / rx;
        double cx = cosPhi * cx1 - sinPhi * cy1 + (p0.X + p1.X) / 2.0;
        double cy = sinPhi * cx1 + cosPhi * cy1 + (p0.Y + p1.Y) / 2.0;

        static double Angle(double ux, double uy, double vx, double vy)
        {
            double length = Math.Sqrt(ux * ux + uy * uy) * Math.Sqrt(vx * vx + vy * vy);
            if (length <= 0) return 0;
            double a = Math.Acos(Math.Min(1, Math.Max(-1, (ux * vx + uy * vy) / length)));
            if (ux * vy - uy * vx < 0) a = -a;
            return a;
        }

        double ux = (x1 - cx1) / rx, uy = (y1 - cy1) / ry;
        double vx = (-x1 - cx1) / rx, vy = (-y1 - cy1) / ry;
        double theta = Angle(1, 0, ux, uy);
        double sweepAngle = Angle(ux, uy, vx, vy);
        if (!sweep && sweepAngle > 0) sweepAngle -= 2 * Math.PI;
        if (sweep && sweepAngle < 0) sweepAngle += 2 * Math.PI;

        // A cubic approximates at most a quarter turn well; split beyond that.
        int segments = Math.Max(1, (int)Math.Ceiling(Math.Abs(sweepAngle) / (Math.PI / 2)));
        double step = sweepAngle / segments;
        double k = 4.0 / 3.0 * Math.Tan(step / 4);
        PointF from = p0;

        for (int s = 0; s < segments; s++)
        {
            double next = theta + step;
            PointF On(double t) => new(
                (float)(cx + rx * cosPhi * Math.Cos(t) - ry * sinPhi * Math.Sin(t)),
                (float)(cy + rx * sinPhi * Math.Cos(t) + ry * cosPhi * Math.Sin(t)));
            PointF Tangent(double t) => new(
                (float)(-rx * cosPhi * Math.Sin(t) - ry * sinPhi * Math.Cos(t)),
                (float)(-rx * sinPhi * Math.Sin(t) + ry * cosPhi * Math.Cos(t)));

            PointF end = On(next), d0 = Tangent(theta), d1 = Tangent(next);
            path.AddBezier(
                from,
                new PointF((float)(from.X + k * d0.X), (float)(from.Y + k * d0.Y)),
                new PointF((float)(end.X - k * d1.X), (float)(end.Y - k * d1.Y)),
                end);
            from = end; theta = next;
        }
    }
}

File: SunoFlowApp/Sources/SunoFlow/BrandMark.swift
import AppKit

/// The SunoFlow brand mark — an ear receiving sound.
///
/// The path strings below are copied verbatim from the website's wordmark
/// (`site/assets/favicon.svg`, and the `<svg>` inside every page's `.wordmark`),
/// so the menu bar, the app icon and the web app cannot drift apart. They are
/// SVG path data in a 24x24 box, stroked at width 2 with round caps and joins.
///
/// Nothing here depends on the rest of the app, so `tools/make-icon.swift` can
/// be compiled against this file to render the `.icns` from the same geometry.
enum BrandMark {

    // MARK: - The mark

    /// The outer ear. A closed subpath, so it can also be filled as a silhouette.
    static let earPath = "M8.9 4.3c2.8 0 4.7 2 4.7 4.8 0 2.5-1.8 3.7-2.9 5.1-.7.9-.9 1.8-.9 2.9 0 1.5-1.2 2.6-2.6 2.6s-2.6-1.2-2.6-2.6c0-1.2.4-2 .4-3.1 0-1.6-1-2.8-1-4.9 0-2.8 2.1-4.8 4.9-4.8z"

    /// The inner curl of the ear.
    static let curlPath = "M8.9 7.4c1.4 0 2.2 1 2.2 2.1 0 1.3-1.1 1.9-1.7 2.8-.4.6-.5 1.2-.5 1.9"

    /// The two arcs of sound arriving at the ear.
    static let wavePaths = [
        "M16.6 10.2a3.6 3.6 0 0 0 0 5.2",
        "M19.9 8.2a6.8 6.8 0 0 0 0 9.2",
    ]

    /// The design box the paths live in, and the weight they are drawn at.
    static let box: CGFloat = 24
    static let stroke: CGFloat = 2

    // MARK: - Variants

    /// How much of the mark to draw.
    ///
    /// The ear is always the mark; what changes is whether sound is arriving
    /// (`idle`), being taken in (`recording` — a filled silhouette, the way SF
    /// Symbols pair `mic` with `mic.fill`), or not being listened for at all
    /// (`processing` is busy, `offline` is struck through).
    enum Variant {
        case idle, recording, processing, offline

        var showsWaves: Bool { self == .idle || self == .recording }
        var isFilled: Bool { self == .recording }
        var isSlashed: Bool { self == .offline }

        var accessibilityDescription: String {
            switch self {
            case .idle: return "SunoFlow — ready"
            case .recording: return "SunoFlow — recording"
            case .processing: return "SunoFlow — transcribing"
            case .offline: return "SunoFlow — speech engine offline"
            }
        }
    }

    // MARK: - Drawing

    /// Draws the mark centred in `rect`, scaled to the shorter side.
    static func draw(_ variant: Variant, in rect: CGRect, color: CGColor, into ctx: CGContext) {
        let scale = min(rect.width, rect.height) / box
        let transform = CGAffineTransform(
            translationX: rect.midX - box * scale / 2,
            y: rect.midY + box * scale / 2
        ).scaledBy(x: scale, y: -scale)   // SVG's y grows downwards; Quartz's grows up.
        let width = stroke * scale

        ctx.saveGState()
        defer { ctx.restoreGState() }

        ctx.setLineCap(.round)
        ctx.setLineJoin(.round)
        ctx.setLineWidth(width)
        ctx.setStrokeColor(color)
        ctx.setFillColor(color)

        let ear = placed(earPath, transform)
        if variant.isFilled {
            // Fill *and* stroke, so the silhouette keeps the same outer edge as
            // the outlined variants and the mark doesn't jump size on record.
            ctx.addPath(ear); ctx.fillPath()
            ctx.addPath(ear); ctx.strokePath()
        } else {
            ctx.addPath(ear); ctx.strokePath()
            ctx.addPath(placed(curlPath, transform)); ctx.strokePath()
        }

        if variant.showsWaves {
            for wave in wavePaths {
                ctx.addPath(placed(wave, transform)); ctx.strokePath()
            }
        }

        if variant.isSlashed {
            let slash = CGMutablePath()
            slash.move(to: CGPoint(x: 4.2, y: 19.8).applying(transform))
            slash.addLine(to: CGPoint(x: 19.8, y: 4.2).applying(transform))
            // Knock a gap out from under the slash first, so it reads as one
            // stroke crossing the mark rather than a line tangled up in it.
            ctx.saveGState()
            ctx.setBlendMode(.clear)
            ctx.setLineWidth(width * 2.4)
            ctx.addPath(slash); ctx.strokePath()
            ctx.restoreGState()
            ctx.addPath(slash); ctx.strokePath()
        }
    }

    /// The mark as a template image, so macOS handles light/dark menu bars and
    /// the pressed-state inversion on its own, and SwiftUI can tint it with
    /// `.foregroundStyle`.
    static func image(_ variant: Variant = .idle, size: CGFloat = 18) -> NSImage {
        let image = NSImage(size: NSSize(width: size, height: size), flipped: false) { rect in
            guard let ctx = NSGraphicsContext.current?.cgContext else { return false }
            // Inset by half the stroke so round caps aren't clipped at the edge.
            let inset = stroke * (size / box) / 2
            draw(variant, in: rect.insetBy(dx: inset, dy: inset),
                 color: NSColor.black.cgColor, into: ctx)
            return true
        }
        image.isTemplate = true
        image.accessibilityDescription = variant.accessibilityDescription
        return image
    }

    // MARK: - SVG path reading

    /// Parsed once; the mark is fixed at build time.
    private static let parsed: [String: CGPath] = {
        var map: [String: CGPath] = [:]
        for d in [earPath, curlPath] + wavePaths { map[d] = parse(d) }
        return map
    }()

    private static func placed(_ d: String, _ transform: CGAffineTransform) -> CGPath {
        let out = CGMutablePath()
        out.addPath(parsed[d] ?? parse(d), transform: transform)
        return out
    }

    /// A deliberately small SVG path reader: it understands exactly the commands
    /// the mark uses (M, L, C, S, A, Z, and their relative forms). The
    /// alternative is hand-porting the curves into Swift, where they would
    /// quietly drift away from the ones the website ships.
    private static func parse(_ d: String) -> CGPath {
        let path = CGMutablePath()
        var scanner = PathScanner(d)
        var current = CGPoint.zero
        var subpathStart = CGPoint.zero
        var lastControl: CGPoint?
        var command: Character = "M"

        while !scanner.isAtEnd {
            if let letter = scanner.takeCommand() { command = letter }
            let relative = command.isLowercase

            func point(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
                relative ? CGPoint(x: current.x + x, y: current.y + y) : CGPoint(x: x, y: y)
            }

            switch Character(command.lowercased()) {
            case "m":
                let p = point(scanner.number(), scanner.number())
                path.move(to: p)
                current = p; subpathStart = p; lastControl = nil
                // Extra coordinate pairs after a moveto are implicit linetos.
                command = relative ? "l" : "L"

            case "l":
                let p = point(scanner.number(), scanner.number())
                path.addLine(to: p); current = p; lastControl = nil

            case "c":
                let c1 = point(scanner.number(), scanner.number())
                let c2 = point(scanner.number(), scanner.number())
                let end = point(scanner.number(), scanner.number())
                path.addCurve(to: end, control1: c1, control2: c2)
                current = end; lastControl = c2

            case "s":
                // Smooth curve: the first control point mirrors the previous one.
                let previous = lastControl ?? current
                let c1 = CGPoint(x: 2 * current.x - previous.x, y: 2 * current.y - previous.y)
                let c2 = point(scanner.number(), scanner.number())
                let end = point(scanner.number(), scanner.number())
                path.addCurve(to: end, control1: c1, control2: c2)
                current = end; lastControl = c2

            case "a":
                let rx = scanner.number(), ry = scanner.number()
                let rotation = scanner.number()
                let largeArc = scanner.number() != 0, sweep = scanner.number() != 0
                let end = point(scanner.number(), scanner.number())
                addArc(to: path, from: current, to: end, rx: rx, ry: ry,
                       rotation: rotation, largeArc: largeArc, sweep: sweep)
                current = end; lastControl = nil

            case "z":
                path.closeSubpath()
                current = subpathStart; lastControl = nil

            default:
                return path   // An unsupported command: stop rather than guess.
            }
        }
        return path
    }

    /// Endpoint-parameterised elliptical arc to cubic Béziers (SVG spec F.6.5).
    private static func addArc(to path: CGMutablePath, from p0: CGPoint, to p1: CGPoint,
                               rx rxIn: CGFloat, ry ryIn: CGFloat, rotation degrees: CGFloat,
                               largeArc: Bool, sweep: Bool) {
        guard p0 != p1 else { return }
        var rx = abs(rxIn), ry = abs(ryIn)
        guard rx > 0, ry > 0 else { path.addLine(to: p1); return }

        let phi = degrees * .pi / 180
        let cosPhi = cos(phi), sinPhi = sin(phi)
        let dx = (p0.x - p1.x) / 2, dy = (p0.y - p1.y) / 2
        let x1 = cosPhi * dx + sinPhi * dy
        let y1 = -sinPhi * dx + cosPhi * dy

        // Grow the radii if they're too small to span the chord (spec F.6.6).
        let lambda = (x1 * x1) / (rx * rx) + (y1 * y1) / (ry * ry)
        if lambda > 1 { let s = sqrt(lambda); rx *= s; ry *= s }

        let sign: CGFloat = (largeArc != sweep) ? 1 : -1
        let numerator = max(0, rx * rx * ry * ry - rx * rx * y1 * y1 - ry * ry * x1 * x1)
        let denominator = rx * rx * y1 * y1 + ry * ry * x1 * x1
        let coefficient = denominator == 0 ? 0 : sign * sqrt(numerator / denominator)
        let cx1 = coefficient * rx * y1 / ry
        let cy1 = -coefficient * ry * x1 / rx
        let cx = cosPhi * cx1 - sinPhi * cy1 + (p0.x + p1.x) / 2
        let cy = sinPhi * cx1 + cosPhi * cy1 + (p0.y + p1.y) / 2

        func angle(_ ux: CGFloat, _ uy: CGFloat, _ vx: CGFloat, _ vy: CGFloat) -> CGFloat {
            let length = sqrt(ux * ux + uy * uy) * sqrt(vx * vx + vy * vy)
            guard length > 0 else { return 0 }
            var a = acos(min(1, max(-1, (ux * vx + uy * vy) / length)))
            if ux * vy - uy * vx < 0 { a = -a }
            return a
        }

        let ux = (x1 - cx1) / rx, uy = (y1 - cy1) / ry
        let vx = (-x1 - cx1) / rx, vy = (-y1 - cy1) / ry
        var theta = angle(1, 0, ux, uy)
        var sweepAngle = angle(ux, uy, vx, vy)
        if !sweep && sweepAngle > 0 { sweepAngle -= 2 * .pi }
        if sweep && sweepAngle < 0 { sweepAngle += 2 * .pi }

        // A cubic approximates at most a quarter turn well; split beyond that.
        let segments = max(1, Int(ceil(abs(sweepAngle) / (.pi / 2))))
        let step = sweepAngle / CGFloat(segments)
        let k = 4.0 / 3.0 * tan(step / 4)
        var from = p0

        for _ in 0..<segments {
            let next = theta + step
            func on(_ t: CGFloat) -> CGPoint {
                CGPoint(x: cx + rx * cosPhi * cos(t) - ry * sinPhi * sin(t),
                        y: cy + rx * sinPhi * cos(t) + ry * cosPhi * sin(t))
            }
            func tangent(_ t: CGFloat) -> CGPoint {
                CGPoint(x: -rx * cosPhi * sin(t) - ry * sinPhi * cos(t),
                        y: -rx * sinPhi * sin(t) + ry * cosPhi * cos(t))
            }
            let end = on(next), d0 = tangent(theta), d1 = tangent(next)
            path.addCurve(to: end,
                          control1: CGPoint(x: from.x + k * d0.x, y: from.y + k * d0.y),
                          control2: CGPoint(x: end.x - k * d1.x, y: end.y - k * d1.y))
            from = end; theta = next
        }
    }
}

/// Character-by-character reader for SVG path data, where numbers may run
/// together without separators (`.4-3.1` is two numbers, `0.4` and `-3.1`).
private struct PathScanner {
    private let characters: [Character]
    private var index = 0

    init(_ string: String) { characters = Array(string) }

    var isAtEnd: Bool {
        var i = index
        while i < characters.count, isSeparator(characters[i]) { i += 1 }
        return i >= characters.count
    }

    private func isSeparator(_ c: Character) -> Bool {
        c == " " || c == "," || c == "\n" || c == "\t" || c == "\r"
    }

    private mutating func skipSeparators() {
        while index < characters.count, isSeparator(characters[index]) { index += 1 }
    }

    /// Consumes and returns a command letter, or nil if the next token is a
    /// number (a repeated coordinate set for the command already in effect).
    mutating func takeCommand() -> Character? {
        skipSeparators()
        guard index < characters.count, characters[index].isLetter else { return nil }
        defer { index += 1 }
        return characters[index]
    }

    mutating func number() -> CGFloat {
        skipSeparators()
        var text = ""
        if index < characters.count, characters[index] == "-" || characters[index] == "+" {
            text.append(characters[index]); index += 1
        }
        var seenDot = false
        while index < characters.count {
            let c = characters[index]
            if c.isNumber {
                text.append(c); index += 1
            } else if c == "." && !seenDot {
                seenDot = true; text.append(c); index += 1
            } else {
                break
            }
        }
        return CGFloat(Double(text) ?? 0)
    }
}
